account_asset_management/models/account_move.py

```diff
@@ -0,0 +1,158 @@
+                    _("You are not allowed to remove an accounting entry "
+                      "linked to an asset."
+                      "\nYou should remove such entries from the asset."))
+            # trigger store function
+            deprs.write({'move_id': False})
+        return super(AccountMove, self).unlink(**kwargs)
+
+    @api.multi
+    def write(self, vals):
+        if set(vals).intersection(FIELDS_AFFECTS_ASSET_MOVE):
+            for move in self:
+                deprs = self.env['account.asset.line'].search(
+                    [('move_id', '=', move.id), ('type', '=', 'depreciate')])
+                if deprs:
+                    raise UserError(
+                        _("You cannot change an accounting entry "
+                          "linked to an asset depreciation line."))
+        return super(AccountMove, self).write(vals)
+
+
+class AccountMoveLine(models.Model):
+    _inherit = 'account.move.line'
+
+    asset_profile_id = fields.Many2one(
+        comodel_name='account.asset.profile',
+        string='Asset Profile')
+    asset_id = fields.Many2one(
+        comodel_name='account.asset',
+        string='Asset', ondelete='restrict')
+
+    @api.onchange('account_id')
+    def _onchange_account_id(self):
+        self.asset_profile_id = self.account_id.asset_profile_id
+
+    @api.model
+    def create(self, vals, **kwargs):
+        if vals.get('asset_id') and not self._context.get('allow_asset'):
+            raise UserError(
+                _("You are not allowed to link "
+                  "an accounting entry to an asset."
+                  "\nYou should generate such entries from the asset."))
+        if vals.get('asset_profile_id'):
+            # create asset
+            asset_obj = self.env['account.asset']
+            move = self.env['account.move'].browse(vals['move_id'])
+            depreciation_base = vals['debit'] or -vals['credit']
+            temp_vals = {
+                'name': vals['name'],
+                'profile_id': vals['asset_profile_id'],
+                'purchase_value': depreciation_base,
+                'partner_id': vals['partner_id'],
+                'date_start': move.date,
+            }
+            if self._context.get('company_id'):
+                temp_vals['company_id'] = self._context['company_id']
+            temp_asset = asset_obj.new(temp_vals)
+            temp_asset._onchange_profile_id()
+            asset_vals = temp_asset._convert_to_write(temp_asset._cache)
+            self._get_asset_analytic_values(vals, asset_vals)
+            ctx = dict(self._context, create_asset_from_move_line=True,
+                       move_id=vals['move_id'])
+            asset = asset_obj.with_context(
+                ctx).create(asset_vals)
+            vals['asset_id'] = asset.id
+        return super(AccountMoveLine, self).create(vals, **kwargs)
+
+    @api.multi
+    def write(self, vals, **kwargs):
+        for aml in self:
```

Comment:
no need to loop if you don't need to use `aml` value. Just do `if self.mapped('asset_id')` and  check FIELDS_AFFECTS_ASSET_MOVE_LINE **before** everything else.